rialto_airflow/harvest/pubmed.py

```diff
@@ -0,0 +1,196 @@
+MAX_RESULTS = 1000  # the maximum number of pubmed IDs we will get for the query
+SEARCH_PATH = f"/entrez/eutils/esearch.fcgi?db=pubmed&retmode=json&retmax={MAX_RESULTS}"  # this endpoint suppots json
+FETCH_PATH = f"/entrez/eutils/efetch.fcgi?db=pubmed&retmode=xml&retmax={MAX_RESULTS}"  # only xml is supported by this endpoint
+HEADERS = {"User-Agent": "stanford-library-rialto", "Accept": "application/json"}
+
+
+def harvest(snapshot: Snapshot, limit=None) -> Path:
+    """
+    Walk through all the Author ORCIDs and generate publications for them from pubmed.
+    """
+    jsonl_file = snapshot.path / "pubmed.jsonl"
+    count = 0
+    stop = False
+
+    with jsonl_file.open("w") as jsonl_output:
+        with get_session(snapshot.database_name).begin() as select_session:
+            # get all authors that have an ORCID
+            for author in (
+                select_session.query(Author).where(Author.orcid.is_not(None)).all()  # type: ignore
+            ):
+                if stop is True:
+                    logging.info(f"Reached limit of {limit} publications stopping")
+                    break
+
+                pmids = pmids_from_orcid(author.orcid)
+                if not pmids:
+                    logging.info(f"No publications found for {author.orcid}")
+                    continue
+
+                for pubmed_pub in publications_from_pmids(pmids):
+                    count += 1
+                    if limit is not None and count > limit:
+                        stop = True
+                        break
+
+                    doi = normalize_doi(get_doi(pubmed_pub))
+
+                    with get_session(snapshot.database_name).begin() as insert_session:
+                        # if there's a DOI constraint violation we need to update instead of insert
+                        pub_id = insert_session.execute(
+                            insert(Publication)
+                            .values(
+                                doi=doi,
+                                pubmed_json=pubmed_pub,
+                            )
+                            .on_conflict_do_update(
+                                constraint="publication_doi_key",
+                                set_=dict(pubmed_json=pubmed_pub),
+                            )
+                            .returning(Publication.id)
+                        ).scalar_one()
+
+                        # a constraint violation is ok here, since it means we
+                        # already know that the publication is by the author
+                        insert_session.execute(
+                            insert(pub_author_association)
+                            .values(publication_id=pub_id, author_id=author.id)
+                            .on_conflict_do_nothing()
+                        )
+
+                        jsonl_output.write(json.dumps(pubmed_pub) + "\n")
+
+    return jsonl_file
+
+
+def pmids_from_orcid(orcid):
```

Comment:
We haven't been consistent about this, but to help with type-checking it might be good to add type annotations for the parameters. So like:

```python
def pmids_from_orcid(orcid: str) -> list[str]:
    ...
```

If you decide to do this, it would be appropriate for other functions in this file as well.